### ISSUE TITLE
Compress lists of points

### ISSUE THREAD

**Issue by julianandrews** [julianandrews/sgf-parse#4]
Right now lists of points never output compressed. If I can come up with a good standard algorithm for compressing point lists that would be nice.

**Comment by julianandrews** [comment]
I've found what looks to be a promising algorithm: https://arxiv.org/pdf/0908.3916v1.pdf. It's surely overkill for this goal, but I may go ahead and implement it because I'm compulsive.

**Comment by julianandrews** [comment]
Clearer write-up here: https://stackoverflow.com/a/6634668/252723

Looking it over, this would be a lot of work for not much benefit. A simple sweep algorithm would handle most common cases well enough.

**Comment by kasbah** [comment]
Thanks for this parser! I am using it for [a WIP pattern search ](https://patterns.calmgo.games/). 

I don't know how useful this is since I did define my own struct / datatype  and it's for serialization but I pack parsed SGFs into a binary that I include in a WASM compilation using a simple bit-packing approach, serde and message pack. 

```rust

pub struct Point {
    pub x: u8,
    pub y: u8,
}

pub enum Color {
    Black,
    White,
}

pub struct Placement {
    pub color: Color,
    pub point: Point,
}

```

I pack these [like so](https://github.com/kasbah/calm-go-pattern-search/blob/7b37f80fedafae726b9248524d2a28774101b41d/rust/common/src/baduk.rs#L205-L234). Over 100,000 games were about 30MB at one point before I started including metadata.

**Comment by julianandrews** [comment]
This issue isn't about compressing the SGF file itself. It's about supporting the SGF spec "compressed point lists" that let you use rectangles to specify a large number of points in a block efficiently: https://www.red-bean.com/sgf/sgf4.html#3.5.1

`sgf-parse` can parse a compressed point list, but will never output a compressed point list when serializing. Ideally, we'd analyze any list of points during serialization and split it into rectangles. It's a solved problem, but doing it correctly is surprisingly involved for a feature that I'm not sure anyone cares about so it's been pretty low on my list of priorities to implement this.

**Comment by yahya-tamur** [comment]
Hello,
I'm writing a program to study go problems, and I thought I should use the VW prop to restrict the view to small portion of the board, which I think is supposed to be done with a compressed point list, see https://www.red-bean.com/sgf/DD_VW.html. 
While compressing a complex collection of points might be difficult and niche, it seems like a common use case and pretty easy to check if the area is a rectangle, and use a single compressed list if that's the case.
I'm planning on writing a different to_string implementation for my use case ( I'd rather not attach  a ~361 length list of coordinates to every problem, in say, a problem set), but I'd be happy to add this behavior to HashSet::to_sgf and send a pull request.

**Comment by julianandrews** [comment]
Hi!

I think just modifying `HashSet::to_sgf` isn't the right solution. "AR", "LN", and "LB" all also go through this code path and shouldn't have compression. My first thought is that instead of `HashSet` the code should be using newtypes so that the different properties can have different implementations. I'm not sure how using newtypes would affect the ergonomics of accessing points, so it's not totally clear if this is a good solution. Maybe instead this should be fixed in `serialize_prop_value`? 

I also wonder if there should maybe be a flag to `serialize` to specify compressed point list handling. The SGF spec itself doesn't specify how compressed point lists should be generated, so there isn't a single "correct" way to compress a list of points, which makes me hesitant to have it be the default behavior with no way to override it.

So, all this is to say that I'm absolutely open to PRs if you want to put in the work to do the design. And I will happily go back and forth on design questions until we're happy with it. But the main thing stopping me from implementing this before now isn't needing code to identify a rectangle!

**Comment by yahya-tamur** [comment]
Yes, I tried to make the change to the code (and also to change SZ to print a single number when possible), and saw the same issue. A newtype might not be a bad solution, especially since there are already newtypes for Point and Text. There's also apparently a nightly feature that can implement a trait for a (u8, u8) and also for generic (A, B). So, something similar might apply to HashSet\<Point\> and HashSet\<P\>. 
But I didn't notice serialization went through serialize_prop_value, so at least the SZ property could be intercepted there.
What about a PointSetToSgf trait, so in order to implement sgf_prop!, you have to provide a way to serialize point collections, similar to how it also requires a Point type which implements ToSgf. In theory, some games supoorted by sgf such as backgammon don't have grids, so it makes sense to me that the rectangle logic should be under go.
